src/buildout/wheel/__init__.py

```diff
@@ -1,56 +1,230 @@
-import distlib.wheel
-import humpty
+import glob
+import logging
 import os
+import os.path
+import shutil
+import sys
+import tempfile
 import pkg_resources
 import setuptools.package_index
-from six.moves.urllib import parse
 import wheel.install
+import wheel.decorator
 import pip.pep425tags
 import zc.buildout.easy_install
+import distutils.command.install
 
-orig_interpret_distro_name = setuptools.package_index.interpret_distro_name
+
+logger = logging.getLogger(__name__)
+
+NAMESPACE_STUB_PATH = os.path.join(os.path.dirname(__file__),
+                                   'namespace_stub.py')
+assert os.path.isfile(NAMESPACE_STUB_PATH)
+
+
+orig_distros_for_location = setuptools.package_index.distros_for_location
 original_wheel_to_egg = zc.buildout.easy_install.wheel_to_egg
+orig_Installer = zc.buildout.easy_install.Installer
+
+
+class Installer(orig_Installer):
+    """
+    zc.buildout.easy_install.Installer class that:
+
+     * feeds all dists in `eggs-directory` to `pkg_resources.Environment`, not
+       just the eggs that `Environment` would auto-detect
+
+     * knows how to install `wheels`.
+
+     * treats all dists inside `eggs-directory` as eggs, even if they're, say,
+       unpacked distros from wheels.
+    """
+
+    def _get_full_path(self):
+        """ Prepend all dists found in self._dest to self._path """
+        # TODO: This method should go into zc.buildout proper.
+        return self._get_dest_dist_paths() + self._path
+
+    def _get_dest_dist_paths(self):
+        # TODO: This method should go into zc.buildout proper.
+        dest = self._dest
+        if dest is None:
+            return []
+        eggs = glob.glob(os.path.join(dest, '*.egg'))
+        dists = [os.path.dirname(dist_info) for dist_info in
+                 glob.glob(os.path.join(dest, '*', '*.dist-info'))]
+        # sort them like pkg_resources.find_on_path() would
+        return pkg_resources._by_version_descending(set(eggs + dists))
+
+    def _env_rescan_dest(self):
+        # TODO: This method should go into zc.buildout proper.
+        self._env.scan(self._get_dest_dist_paths())
+
+    def _maybe_eggify(self, dist):
+        """ Make sure everything installed under `._dest` is seen as an egg """
+        # TODO: This method and its two uses below can go in buildout proper
+        # without altering its current behavior.
+        if dist.location.startswith(self._dest + '/'):
+            dist.precedence=pkg_resources.EGG_DIST
+        return dist
+
+    def _env_get_best_match(self, requirement, ws):
+        dist = orig_Installer._env_get_best_match(self, requirement, ws)
+        return self._maybe_eggify(dist)
+
+    def _env_get_dists_for(self, project_name):
+        dists = orig_Installer._env_get_dists_for(self, project_name)
+        return [self._maybe_eggify(dist) for dist in dists]
```

Comment:
As mentioned in the review of the corresponding buildout branch, I'd like to go ahead and move these over there.